File: practicas/practica4_rb/lib/Damage.rb
#encoding: utf-8

#Clase que representa el daño producido a una estación espacial por una nave enemiga cuando se pierde un combate

module Deepspace

  class Damage

    private_class_method :new

    def initialize(s)
      @nShields = s
    end

    #Constructor de copia
    def copy()
    end

    def self.newCopy(otro)
      otro.copy
    end

    def discardWeapon(w) end

    def discardShieldBooster
      if nShields > 0
        @nShields -= 1
      else
        @nShields = 0
      end
    end

    def hasNoEffect
      @nShields.zero?
    end

    def adjust(w, s)
    end

    def adjustShields(s)
      [s.length, nShields].min
    end

    def getUIversion
    end

    def to_s
      "nShields: " + @nShields.to_s + "\n"
    end

  end
end

File: practicas/practica5/src/main/java/deepspace/Damage.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

/**
 * Representa el daño producido a una estación espacial por una naveenemiga cuando se pierde un combate
 */
abstract public class Damage {
    private int nShields;
    
    Damage(int s){nShields = s;}
    
    public int getNShields(){return nShields;}
    
    
    
    abstract Damage copy();
    
    abstract DamageToUI getUIversion();
    
  
    
    protected int adjustShields(ArrayList<ShieldBooster> s){
        return Integer.min(s.size(), getNShields());
    }
    
    /**
     * Devuelve una versión ajustada del objeto a las colecciones de armas y potenciadores de escudos suministradas como parámetro.
     * @param w armas
     * @param s potenciadores
     * @return version ajustada del objeto
     */
    abstract public Damage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s);
    
    
    
    /**
     * intenta eliminar el tipo del arma pasada como parámetro de esa lista.
     * En otro caso simplemente decrementa en una unidad el contador de armas que deben ser eliminadas.
     * Ese contador no puedeser inferior a cero en ningún caso
     * @param w el arma
     */
    
    abstract public void discardWeapon(Weapon w);
    
    /**
     * Descarta un shieldBooster
     */
    public void discardShieldBooster() {
        nShields--;
        if(nShields < 0){
            nShields = 0;
        }
    }
    
    /**
     * @return true si el daño representado no tiene ningún efecto.
     * Esto quiere decir que no implica la pérdida de ningún tipo de accesorio (armas o potenciadores de escudo).
     */
    public boolean hasNoEffect(){
        return getNShields() == 0;
    }


    @Override
    public String toString(){
        return "nShields: " + getNShields();
    }
    
    
    
}
